use code_blockt::add and ::move for adding a statement

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

This is the trivial part of #2830.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
use code_blockt::add and code_blockt constructor for adding a statement

.move_operands() is avoided; there is no measurable performance benefit.

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2744,7 +2744,7 @@ codet java_bytecode_convert_methodt::convert_iinc(
   code_assign.rhs() =
     plus_exprt(variable(arg0, 'i', address, CAST_AS_NEEDED), arg1_int_type);
-  block.copy_to_operands(code_assign);
+  block.add(code_assign);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -215,13 +215,13 @@ exprt allocate_dynamic_object(
     code_assignt assign=code_assignt(malloc_sym.symbol_expr(), malloc_expr);
     assign.add_source_location()=loc;
-    output_code.copy_to_operands(assign);
+    output_code.add(assign);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -215,13 +215,13 @@ exprt allocate_dynamic_object(
     code_assignt code(target_expr, malloc_symbol_expr);
     code.add_source_location()=loc;
-    output_code.copy_to_operands(code);
+    output_code.add(code);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -230,7 +230,7 @@ exprt allocate_dynamic_object(
     code_assignt code(target_expr, null_pointer_expr);
     code.add_source_location()=loc;
-    output_code.copy_to_operands(code);
+    output_code.add(code);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -320,7 +320,7 @@ exprt java_object_factoryt::allocate_object(
       code_assignt code(target_expr, aoe);
       code.add_source_location()=loc;
-      assignments.copy_to_operands(code);
+      assignments.add(code);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1079,7 +1077,7 @@ void java_object_factoryt::gen_nondet_struct_init(
       code_assignt code(me, from_integer(0, me.type()));
       code.add_source_location() = loc;
-      assignments.copy_to_operands(code);
+      assignments.add(code);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1237,7 +1235,7 @@ void java_object_factoryt::gen_nondet_init(
     assign.add_source_location()=loc;
 
-    assignments.copy_to_operands(assign);
+    assignments.add(assign);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1408,13 +1406,13 @@ void java_object_factoryt::gen_nondet_array_init(
 
   exprt java_zero=from_integer(0, java_int_type());
-  assignments.copy_to_operands(code_assignt(counter_expr, java_zero));
+  assignments.add(code_assignt(counter_expr, java_zero));
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1393,7 +1391,7 @@ void java_object_factoryt::gen_nondet_array_init(
   codet data_assign=code_assignt(array_init_symexpr, init_array_expr);
   data_assign.add_source_location()=loc;
-  assignments.copy_to_operands(data_assign);
+  assignments.add(data_assign);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1328,7 +1326,7 @@ void java_object_factoryt::allocate_nondet_length_array(
   code_assignt assign(lhs, java_new_array);
   assign.add_source_location() = loc;
-  assignments.copy_to_operands(assign);
+  assignments.add(assign);
```
move?

peterschrammel commented on jbmc/src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -170,7 +170,7 @@ void java_simple_method_stubst::create_method_stub(symbolt &symbol)
       init_symbol_expression, symbol_exprt(this_argument.get_identifier()));
     get_argument.add_source_location() = synthesized_source_location;
-    new_instructions.copy_to_operands(get_argument);
+    new_instructions.add(get_argument);
```
move?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 29bd494dc2eda7c5b56f5775c8c1ac03e87e7b58).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86391063

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -141,7 +141,7 @@ static void monitor_exits(codet &code, const codet &monitorexit)
     return_block.add(monitorexit);
-    return_block.move_to_operands(code);
+    return_block.move(code);
     code = return_block;
```
How about `code = code_blockt({monitorexit, code});`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -262,11 +262,11 @@ static void instrument_synchronized_code(
+  try_block.move(code);
+  try_block.move(catch_pop);
+  try_block.move(catch_label);
   code = try_block;
```
`code = code_blockt({monitorenter, catch_push, code, catch_pop, catch_label});`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -856,7 +856,7 @@ void java_bytecode_convert_classt::add_array_types(symbol_tablet &symbol_table)
 
     code_declt declare_cloned(local_symexpr);
-    clone_body.move_to_operands(declare_cloned);
+    clone_body.move(declare_cloned);
```
Move the declaration of `clone_body` further down and do `code_blockt clone_body({declare_cloned, create_blank, declare_index, copy_loop, return_inst});` (if I got this right from what I see in the code review).

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2258,8 +2258,8 @@ void java_bytecode_convert_methodt::convert_invoke(
-      ret_block.move_to_operands(clinit_call);
-      ret_block.move_to_operands(c);
+      ret_block.move(clinit_call);
+      ret_block.move(c);
```
`code_blockt ret_block({clinit_call, c});`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2325,8 +2325,8 @@ void java_bytecode_convert_methodt::convert_athrow(
-    ret_block.move_to_operands(assume_code);
+    ret_block.move(assert_code);
+    ret_block.move(assume_code);
     c = ret_block;
```
`code_blockt ret_block({assert_code, assume_code});` (and other places use `std::move` afterwards).

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2399,8 +2399,8 @@ codet &java_bytecode_convert_methodt::do_exception_handling(
-      try_block.move_to_operands(catch_push);
-      try_block.move_to_operands(c);
+      try_block.move(catch_push);
+      try_block.move(c);
```
`code_blockt try_block({catch_push, c});`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2440,8 +2440,8 @@ codet &java_bytecode_convert_methodt::do_exception_handling(
-        end_try_block.move_to_operands(c);
-        end_try_block.move_to_operands(catch_pop);
+        end_try_block.move(c);
+        end_try_block.move(catch_pop);
```
`code_blockt end_try_block({c, catch_pop});`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2552,8 +2552,8 @@ void java_bytecode_convert_methodt::convert_new(
-    ret_block.move_to_operands(clinit_call);
-    ret_block.move_to_operands(c);
+    ret_block.move(clinit_call);
+    ret_block.move(c);
```
`code_blockt ret_block({clinit_call, c});`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -129,8 +129,8 @@ codet java_bytecode_instrumentt::throw_exception(
-  throw_code.move_to_operands(assign_new);
-  throw_code.copy_to_operands(code_expressiont(throw_expr));
+  throw_code.move(assign_new);
+  throw_code.add(code_expressiont(throw_expr));
```
`code_blockt throw_code({assign_new, code_expressiont(throw_expr)});`

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -481,9 +481,9 @@ void c_typecheck_baset::typecheck_for(codet &code)
     code.op0().make_nil();
-    code_block.move_to_operands(code);
+    code_block.move(code);
     code.swap(code_block);
```
This code is a bit strange: the `code.op0().make_nil();` appears to have lasting effect.

kroening replied:
yes, see comment above the if(...) -- this is deliberate.

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -627,7 +627,7 @@ void c_typecheck_baset::typecheck_ifthenelse(code_ifthenelset &code)
     code_block.add_source_location()=code.then_case().source_location();
 
-    code_block.move_to_operands(code.then_case());
+    code_block.move(code.then_case());
```
`code_blockt code_block({code.then_case()});`

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -640,7 +640,7 @@ void c_typecheck_baset::typecheck_ifthenelse(code_ifthenelset &code)
       code_block.add_source_location()=code.else_case().source_location();
 
-      code_block.move_to_operands(code.else_case());
+      code_block.move(code.else_case());
```
`code_blockt code_block(code.else_case());`

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -760,7 +760,7 @@ void c_typecheck_baset::typecheck_while(code_whilet &code)
     code_block.add_source_location()=code.body().source_location();
 
-    code_block.move_to_operands(code.body());
+    code_block.move(code.body());
```
`code_blockt code_block({code.body});`

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -795,7 +795,7 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
     code_block.add_source_location()=code.body().source_location();
 
-    code_block.move_to_operands(code.body());
+    code_block.move(code.body());
```
`code_blockt code_block({code.body()});`

tautschnig commented on src/cpp/cpp_typecheck.cpp:
```diff
@@ -177,7 +177,7 @@ void cpp_typecheckt::static_and_dynamic_initialization()
       // which we execute.
       assert(symbol.value.id()==ID_code);
-      init_block.copy_to_operands(symbol.value);
+      init_block.add(to_code(symbol.value));
```
The `assert` just above can be dropped now.

tautschnig commented on src/cpp/cpp_typecheck_code.cpp:
```diff
@@ -163,7 +163,8 @@ void cpp_typecheckt::typecheck_switch(code_switcht &code)
     code_blockt code_block;
-    code_block.move_to_operands(decl.op0(), code);
+    code_block.move(to_code(decl.op0()));
+    code_block.move(code);
```
`code_blockt code_block({to_code(decl.op0()), code});`

tautschnig commented on src/goto-instrument/goto_program2code.cpp:
```diff
@@ -1748,7 +1748,7 @@ void goto_program2codet::cleanup_code_ifthenelse(
       // ambiguity
       code_blockt b;
-      b.move_to_operands(i_t_e.then_case());
+      b.move(i_t_e.then_case());
```
`code_blockt b({i_t_e.then_case()});`

tautschnig commented on src/goto-instrument/goto_program2code.cpp:
```diff
@@ -1760,7 +1760,7 @@ void goto_program2codet::cleanup_code_ifthenelse(
       // ambiguity
       code_blockt b;
-      b.move_to_operands(i_t_e.else_case());
+      b.move(i_t_e.else_case());
```
`code_blockt b({i_t_e.else_case()});`

tautschnig commented on src/goto-instrument/goto_program2code.cpp:
```diff
@@ -1779,9 +1779,9 @@ void goto_program2codet::cleanup_code_ifthenelse(
-      b.move_to_operands(else_label);
+      b.move(i_t_e);
+      b.move(then_label);
+      b.move(else_label);
```
`code_blockt b({i_t_e, then_label, else_label});`

tautschnig commented on src/util/std_code.cpp:
```diff
@@ -123,8 +123,8 @@ code_blockt create_fatal_assertion(
-  result.copy_to_operands(code_assertt(condition));
-  result.copy_to_operands(code_assumet(condition));
+  result.add(code_assertt(condition));
+  result.add(code_assumet(condition));
```
`code_blockt result({code_assertt(condition), code_assumet(condition)});`

kroening commented:
@tautschnig @peterschrammel Please note that your suggestions to into opposite directions: Michael's means more copying, Peter wants less.

tautschnig commented:
> Please note that your suggestions to into opposite directions: Michael's means more copying, Peter wants less.

Well, yes, but on different bits of the code base. @peterschrammel's suggestions exclusively cover cases that require (at least so it seems) incremental construction; mine talks about cases where likely we could even add a `const` as the object is fully constructed in just one line of code.

kroening commented:
Perhaps to make more fundamental progress on this:
```
  for(int i=0; i<2000000; i++)
  {
    irep_idt id;
    symbol_exprt s(id, bool_typet());
    exprt something;
    something.move_to_operands(s);
    //something.copy_to_operands(s);
  }
```
with move_to_operands: 0.59s
with copy_to_operands: 0.55s

This does not appear to require a recent compiler. The difference is actually difficult to measure (it's in the range of the noise). Thus, there is absolutely no reason to believe that moving ireps would be any faster than copying.

The key reason why the moving methods exist is that they did offer enormous benefit before the sharing was introduced. I am thus tempted to remove the .move() method, say for code_blockt as a starter. It's simply a distraction.

tautschnig commented:
@kroening Thank you for doing this measurement! Radical as I am, may I propose the even bigger step of getting rid of `move_to_operands` altogether (in the long term, not in this PR)? And then we could also inline `copy_to_operands`...

kroening commented:
Done; please take another look.

kroening commented:
@tautschnig Indeed, will do a PR that deprecates move_to_operands().

smowton commented:
Hang on, the move operations is surely only similar like this when there is no usage of the moved `irept` subsequently. For example,

```
irept i1 = something();
irept copy = i1;
copy.id() = "hello world";
```

This will cause a CoW break upon calling the non-const variant of `.id()`, whereas if we had moved `i1` there would be no break. That sort of situation would surely produce a significant difference.

In the particular case discussed here we would have something like

```
block.move_to_operands(something);
...
replace_symbol(block, ...);
```

If `something` was copied then `replace_symbol` may cause a CoW break, whereas if it was moved then it won't.

kroening commented:
@smowton Indeed, this assumes that the moved object isn't touched afterwards; in your example, I'd assume that 'something' isn't ever used after the move_to_operands. I would assume that to be the case across the code base, and would consider anything else to be a bug.

smowton commented:
Right, but operations against the destination of the move *are* allowed, and if they take place before the moved-from handle has died then a needless CoW break will occur.

smowton commented:
(in both my examples it's the target that's being altered, and the source handle just happens to still be alive)

kroening commented:
@smowton Indeed, that may have some performance cost.

smowton commented:
So then, surely keep the move ops but note in reviews when they're redundant because the moved-from thing is about to die regardless?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 234ac62f1ea9f97ec046a0db398f75ad62d6da88).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86463616

smowton commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -139,10 +139,7 @@ static void monitor_exits(codet &code, const codet &monitorexit)
-    return_block.add(monitorexit);
-    return_block.move_to_operands(code);
-    code = return_block;
+    code = code_blockt({monitorexit, code});
```
If you're going for this style, could I recommend replacing the constructor for `code_blockt` that takes a `std::list<const codet &>` with one taking a `std::initializer_list`? The former means a bunch of heap allocates and deallocates, whereas the latter is basically a type-safe layer atop varargs.

smowton replied:
(It could also take universal references, permitting `{monitorexit, std::move(code)}`)

smowton replied:
(Correction, it's not a wrapper on `<stdarg>` style varargs as I'd thought, but it is a fixed-size array, probably stack-allocated, which still avoids much of the pointless back-and-forth of a temporary linked list)

smowton replied:
(You could also use a variadic template... but that's probably more fuss than it's worth when the initializers are monovariant)

kroening replied:
Indeed, there is unnecessary copying going on in the constructor. How about these:
```
  explicit code_blockt(const std::vector<codet> &_statements):codet(ID_block)
  {
    operands()=(const std::vector<exprt> &)_statements;
  }

  explicit code_blockt(std::vector<codet> &&_statements):codet(ID_block)
  {
    operands()=std::move((std::vector<exprt> &&)_statements);
  }
```

smowton replied:
Why not `std::initializer_list<codet>`?  Should be just like vector, except without the heap alloc; plus it's (one of the) new standard ways to do varargs

kroening replied:
Does that add anything over offering the vector interface? Note that the vector offers a constructor from std::initializer_list<codet>, and that the vector needs to exist eventually.

smowton replied:
I think marginally, but the difference is small so let's go with this. It's definitely much better than `std::list` for this purpose.

smowton reviewed (commented):
Recommend using `std::initializer_list`, and as previously mentioned, object to `move` -> `add` in general.

kroening committed:
new constructors for code_blockt

The new constructors avoid building a vector from a list.  The vector is
constructed directly at the call site of the constructor.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8c40aa14ca3f8f29513e72d719c4d8fd9c8366ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86610869

kroening commented:
@smowton To make progress on this, how about a variant of code_blockt::add that gets an rvalue reference? You'd then write
```
block.add(std::move(something));
```
in the scenario in which you would modify the added statement after adding.

smowton commented:
Sure? Don't mind re: `block.move(codet)` and `block.add(codet &&)`, I just don't want to needlessly throw away a useful opportunity to reduce CoW breaks

kroening committed:
rvalue variants of exprt::copy_to_operands

kroening committed:
rvalue variant of code_blockt::add

kroening commented:
Ok, done!

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 28beb76741e4a5e7b0e5aa4e8f3dcfc239821ec9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86726683

peterschrammel reviewed (approved):
LGTM modulo clang-format https://travis-ci.org/diffblue/cbmc/jobs/436556422#L509

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 46232f03d3a3176700719149a5d2218b96fd9c78).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87134284
